test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs

```diff
@@ -18,5 +28,185 @@ public void UnwrapAggregateException()
+
+                var args = new DataServiceClientRequestMessageArgs(
+                    "GET",
+                    new Uri("http://localhost"),
+                    useDefaultCredentials: false,
+                    usePostTunneling: false,
+                    new Dictionary<string, string>(),
+                    httpClientHandlerProvider);
+
+                using (var request = new HttpClientRequestMessage(args))
+                {
+                    // Act
+                    IODataResponseMessage response = request.GetResponse();
+                    var stream = response.GetStream();
+                    string contents = "";
+                    using (StreamReader reader = new StreamReader(stream))
+                    {
+                        contents = reader.ReadToEnd();
+                    }
+
+                    // Assert
+                    Assert.Equal(expectedResponse, contents);
+                    Assert.Equal(1, handler.Requests.Count);
+                    Assert.Equal("GET http://localhost/", handler.Requests[0]);
+                    Assert.Equal(1, httpClientHandlerProvider.NumCalls);
+                }
+            }
+        }
+
+        [Fact]
+        public async void WhenHttpClientHandlerIsProvided_DoesNotDisposeHandler()
+        {
+            // Arrange
+            string expectedResponse = "Foo";
+            using (var handler = new MockHttpClientHandler(expectedResponse))
+            {
+
+                var httpClientProvider = new MockHttpClientHandlerProvider(handler);
+
+                var args = new DataServiceClientRequestMessageArgs(
+                    "GET",
+                    new Uri("http://localhost"),
+                    useDefaultCredentials: false,
+                    usePostTunneling: false,
+                    new Dictionary<string, string>(),
+                    httpClientProvider);
+
+                var request = new HttpClientRequestMessage(args);
+
+                // Act
+                IODataResponseMessage response = request.GetResponse();
+                var stream = response.GetStream();
+                stream.Dispose();
+                request.Dispose();
+
+                // Re-use handler with a different HttpClient after initial request is disposed
+                var client = new HttpClient(handler, disposeHandler: false);
+                var newResponse = await client.GetAsync("http://foo");
+
+                // Assert
+                // We can still make request from the client after HttpClientRequestMessage is disposed
+                var contents = await newResponse.Content.ReadAsStringAsync();
+                Assert.Equal(expectedResponse, contents);
+                Assert.Equal(2, handler.Requests.Count);
+                Assert.Equal("GET http://localhost/", handler.Requests[0]);
+                Assert.Equal("GET http://foo/", handler.Requests[1]);
+                Assert.False(handler.Disposed);

```

Comment:
```suggestion
                using (var request = new HttpClientRequestMessage(args))
                {
                    // Act
                    IODataResponseMessage response = request.GetResponse();
                    var stream = response.GetStream();
                    stream.Dispose();
                    request.Dispose();

                    // Re-use handler with a different HttpClient after initial request is disposed
                    using (var client = new HttpClient(handler, disposeHandler: false))
                    using (var newResponse = await client.GetAsync("http://foo"))
                    {
                        // Assert
                        // We can still make request from the client after HttpClientRequestMessage is disposed
                        var contents = await newResponse.Content.ReadAsStringAsync();
                        Assert.Equal(expectedResponse, contents);
                        Assert.Equal(2, handler.Requests.Count);
                        Assert.Equal("GET http://localhost/", handler.Requests[0]);
                        Assert.Equal("GET http://foo/", handler.Requests[1]);
                        Assert.False(handler.Disposed);
                    }
                }
```

Reply:
Here I wanted to make sure the request is disposed *before* creating the new client. If I create a `using` block, maybe it makes sense to close before the line that creates a new client and to do away with the explicit `request.Dispose()` and `stream.Dispose()` calls.